### PR TITLE
[SYCL-PTX] Update documentation for the CUDA backend.

### PR DESCRIPTION
**Pull request by Naghasan**
Sync the post link process for PTX following the illustration update. Add a description for the global offset handling.

Signed-off-by: Victor Lomuller <victor@codeplay.com>

### CODE REVIEW

File: sycl/doc/CompilerAndRuntimeDesign.md [hunk 1]
```diff
@@ -510,15 +510,16 @@ down to the NVPTX Back End. All produced bitcode depends on two libraries,
...
 (built by the libclc project).
 
-During the device linking step (device linker box in the
-[Separate Compilation and Linking](#separate-compilation-and-linking)
-illustration), llvm bitcode objects for the CUDA target are linked together
-alongside `libspirv-nvptx64--nvidiacl.bc` and `libdevice.bc`, compiled to PTX
-using the NVPTX backend, and assembled into a cubin using the `ptxas` tool (part
+##### Device code post-link step
+
+During the "PTX target precessing" in the device linking step [Device code post-link step](#device-code-post-link-step), the llvm bitcode objects for the CUDA target are linked together alongside `libspirv-nvptx64--nvidiacl.bc` and `libdevice.bc`, compiled to PTX
```
Comment (pvchupin) [inline sycl/doc/CompilerAndRuntimeDesign.md:515]:
```suggestion
During the "PTX target processing" in the device linking step [Device code post-link step](#device-code-post-link-step), the llvm bitcode objects for the CUDA target are linked together alongside `libspirv-nvptx64--nvidiacl.bc` and `libdevice.bc`, compiled to PTX
```

Comment (pvchupin) [reply sycl/doc/CompilerAndRuntimeDesign.md:515]: Please format to 80 chars lines.

File: sycl/doc/CompilerAndRuntimeDesign.md [hunk 2]
```diff
@@ -592,9 +593,82 @@ define void @SYCL_generated_kernel(i32 %local_ptr_offset, i32 %arg, i32 %local_p
...
 implementation will internally set the argument as the offset with respect to
-the accumulated size of used local memory. This approach preserves the exisiting
+the accumulated size of used local memory. This approach preserves the existing
 PI interface.
 
+##### Global offset support
+
+The CUDA API does not natively support the global offset parameter expected by the SYCL.
+
+In order to emulate this and make generated kernel compliant, an intrinsic `llvm.nvvm.implicit.offset` (clang builtin `__builtin_ptx_implicit_offset`) was introduce materializing the use of this implicit parameter for the NVPTX backend. The intrinsic return a pointer to `i32` referring to a 3 elements array.
```
Comment (pvchupin) [inline sycl/doc/CompilerAndRuntimeDesign.md:603]: Please format to 80 chars lines

File: sycl/doc/CompilerAndRuntimeDesign.md (hunk 1 above)
Comment (kbobrovs) [reply sycl/doc/CompilerAndRuntimeDesign.md:515]: on the picture I can see "libspirv" and "libdevice.bc" passed to "Clang"
1) Looks like "libspirv" should be "libspirv.bc"
2) Looks like "clang" is really PTX back-end - I suggest to replace "Clang" with "PTX backend" then, similarly to the generic DeviceLinkAndWrap terminology

Also does PTX BE accept multiple bitcode files or there is llvm-link before it? If the latter, I'd suggest to reflect that on the picture.

File: sycl/doc/CompilerAndRuntimeDesign.md (hunk 2 above)
Comment (kbobrovs) [inline sycl/doc/CompilerAndRuntimeDesign.md:603]:
```suggestion
In order to emulate this and make generated kernel compliant, an intrinsic `llvm.nvvm.implicit.offset` (clang builtin `__builtin_ptx_implicit_offset`) was introduced materializing the use of this implicit parameter for the NVPTX backend. The intrinsic returns a pointer to `i32` referring to a 3 elements array.
```

File: sycl/doc/CompilerAndRuntimeDesign.md [hunk 3]
```diff
@@ -592,9 +593,82 @@ define void @SYCL_generated_kernel(i32 %local_ptr_offset, i32 %arg, i32 %local_p
...
+
+Each non-kernel function reaching the implicit offset intrinsic in the call graph is augmented with an extra implicit parameter of type pointer to `i32`.
+Each kernel calling one of these functions using this intrinsic are cloned:
```
Comment (kbobrovs) [inline sycl/doc/CompilerAndRuntimeDesign.md:606]:
```suggestion
Kernels calling one of these functions using this intrinsic are cloned:
```

File: sycl/doc/CompilerAndRuntimeDesign.md [hunk 4]
```diff
@@ -592,9 +593,82 @@ define void @SYCL_generated_kernel(i32 %local_ptr_offset, i32 %arg, i32 %local_p
...
+
+- the original kernel initialize an array of 3 `i32` to 0 and passes the pointer to this array to each function with the implicit parameter;
```
Comment (kbobrovs) [inline sycl/doc/CompilerAndRuntimeDesign.md:608]:
```suggestion
- the original kernel initializes an array of 3 `i32` to 0 and passes the pointer to this array to each function with the implicit parameter;
```

File: sycl/doc/CompilerAndRuntimeDesign.md [hunk 5]
```diff
@@ -592,9 +593,82 @@ define void @SYCL_generated_kernel(i32 %local_ptr_offset, i32 %arg, i32 %local_p
...
+- the cloned function type is augmented with an implicit parameter of type array of 3 `i32`. The pointer to this array is then passed each function with the implicit parameter.
+
+The runtime will query both kernel and call the appropriate one based on the following logic:
```
Comment (kbobrovs) [inline sycl/doc/CompilerAndRuntimeDesign.md:611]:
```suggestion
The runtime will query both kernels and call the appropriate one based on the following logic:
```

File: sycl/doc/CompilerAndRuntimeDesign.md (hunk 1 above)
Comment (Naghasan) [reply sycl/doc/CompilerAndRuntimeDesign.md]: I applied your comments on the picture.

It is llvm-link before, but it is clearly shown in the DeviceLinkAndWrap, the picture is a split of the "PTX target processing" box. I added reminder on the input box that it is linked.

**Comment by Naghasan** [comment]
I applied the comments and formatted to 80 col.

**Pull request merged**